Abins: explicit cache directory management

ajjackson opened the pull request:
### Description of work
Fixes #38084 


#### Summary of work
- Add an input parameter to Abins/Abins2D which explicitly sets the cache file directory. The default value is the Default Save Directory so if the user does nothing files will continue to save in the same place. But they are more likely to notice it!
- Use temporary directories throughout test suite to avoid collisions and simplify cleanup
  - By refactoring this into the "check" function, some boilerplate is removed from loader tests.
- Remove existing (more verbose and spooky) cleanup mechanism from Abins tests

#### Purpose of work

Abins uses the Default Save Directory. This has a few drawbacks:
- Sometimes it is set to a bad location, and not easy to change (mostly addressed in #38606)
- Users may be unaware that they are creating cache files that are never cleaned up
- Parallel tests and workflows may have unintended interactions through a common cache file
- It is not obvious from the code which functions will interact with a cache


<!-- If the original issue was raised by a user they should be named here. Do not leak email addresses
**Report to:** [user name]
-->

#### Further detail of work
- Cache *directory* was chosen rather than cache *filename* to avoid a likely future API-break: I am considering migrating all the caching to use joblib (already a Mantid dependency) and this caches to multiple files in a directory structure.

### To test:
The changes are covered by unit tests and system tests.

For fun, you can try setting the Default Save Directory to somewhere unwriteable and running the tests. Abins tests used to fail in this scenario, but now will pass because they are all directed to temporary cache directories.


- [x] Add release note

<!-- delete this if you added release notes
*This does not require release notes* because **fill in an explanation of why**
If you add release notes please save them as a separate file using the Issue or PR number as the file name. Check the file is located in the correct directory for your note(s).
-->

<!-- Ensure the base of this PR is correct (e.g. release-next or main)
Finally, don't forget to add the appropriate labels, milestones, etc.!  -->

---

### Reviewer

Please comment on the points listed below ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)).
**Your comments will be used as part of the gatekeeper process, so please comment clearly on what you have checked during your review.** If changes are made to the PR during the review process then your final comment will be the most important for gatekeepers. In this comment you should make it clear why any earlier review is still valid, or confirm that all requested changes have been addressed.

#### Code Review

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?
- Do the release notes conform to the [release notes guide](https://developer.mantidproject.org/Standards/ReleaseNotesGuide.html)?

#### Functional Tests

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### Gatekeeper

If you need to request changes to a PR then please add a comment and set the review status to "Request changes". This will stop the PR from showing up in the list for other gatekeepers.

ajjackson committed:
Drop __init__ from AbinsAlgorithm

It doesn't seem to do anything useful

ajjackson committed:
Accept CacheDirectory user input to Abins (currently does nothing)

Add an input parameter "CacheDirectory" which is validated as being
writeable. This gives a nice user experience, with default value
appearing when the field is emptied.

The actual caching mechanics still need to be connected to this
parameter.

ajjackson committed:
Modify Abins test cleanup to take cache dir; update loader tests

Other tests (which create cache in different ways) still fail, this is
expected.

ajjackson committed:
Update more tests for explicit cache dir

ajjackson committed:
Fix Abins Unit tests; revised cleanup function signature

We are also starting to pass the cache directory to Abins explicitly,
although it is not used yet.

ajjackson committed:
Use explicit cache_directory variable in Abins algorithms

ajjackson committed:
Ensure explicit cache_directory is used across Abins and tests

- Temporarily raise an error if the default cache_directory (None) is
  passed to abins.io.IO
- This helped identify where tests were using default dirs. Now there
  are updated to use a (somewhat inconsistent) mixture of temporary
  directories and the user default directory

ajjackson committed:
More use of tmpdir on Abins tests

- With explicit cache paths we can remove the need to patch IO class
  and simplify tests
- Using TemporaryDirectory.cleanup() simplifies post-test clean-up

ajjackson committed:
Rely entirely on tempfile to cleanup Abins tests

ajjackson committed:
More consistent attribute names

ajjackson committed:
Make cache_directory argument mandatory

ajjackson committed:
More consistent attribute names

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

ajjackson commented:
Looks like the doctest does need updating

ajjackson commented:
The doctest failure is related to calling the Algorithm from Python without explicitly including the new CacheDirectory argument, which is apparently mandatory. I have provided a default value and would prefer that it works as an optional argument; is there a way to make this work like the other optional parameters?

ajjackson committed:
Add CacheDirectory to Abins-1 doctest

Abins2D will still break.

I'm not really happy with this "fix" because there _should_ be a
default argument.

ajjackson commented:
Have pushed a partial fix to see if it works. For some reason on my machine doc-tests take forever to run and give no output; cannot tell if frozen or just very slow.

ajjackson committed:
Fix Abins2D doctest (but a bit clunky)

As with Abins-v1, we fix the doctest by
- providing a CacheDirectory argument to the Algorithm calls
- setting a corresponding cleanup path

The problems with this approach are:
- it's ugly and distracting
- it _should_ be unnecessary, there is a default value and it works
when calling from GUI
- /tmp probably doesn't work on Windows

ajjackson commented:
Doctest fixed but could use some further input: as stated in commit message

> As with Abins-v1, we fix the doctest by
> - providing a CacheDirectory argument to the Algorithm calls
> - setting a corresponding cleanup path
> 
> The problems with this approach are:
> - it's ugly and distracting
> - it _should_ be unnecessary, there is a default value and it works
> when calling from GUI
> - /tmp probably doesn't work on Windows

ajjackson committed:
Add release note

adriazalvarez commented:
@ajjackson Changes are looking good, I just have a more general question about the cache files. 
Do the users normally want to check the .hdf files after the output workspaces have been processed? It seems is the job of the user to delete the cache files once abins is done. 
Then, we could set up the temporary files in temporary directories (probably python's tempfile wiill assign the correct temporary folder for every system) and delete them when the algo is done with them (as in the tests)? 
Let me know if I'm missing something.

ajjackson commented:
The idea behind the cache file is that the user might want to run Abins a few times with slightly different settings (e.g. different ISIS instrument, different configuration of that instrument, or difference cross-section weighting). The cache can be used to avoid unnecessary re-computation in that case. As the cache file is named after the input file, they can do this with some persistence across a few input systems and across Mantid sessions.

I think in practice:
- this is less necessary than it used to be because the performance of Abins has improved a lot. But it might still be helpful for some more heroic cases.
- Most Abins users don't know where the file goes so they hang around longer than necessary.
- Most of that benefit could be had by maintaining the cache for a single Mantid session and cleaning up when Mantid closes.
- But in the cases where the user benefits the most, it might be good to keep them.

One middle-ground option could be to set the default cache directory to a system /tmp directory (or whatever the Windows equivalent is) so the cache persists until the next reboot, while "power users" have the option to nominate a user directory and keep them.

Destroying the cache when the Abins Algorithm finishes is generally "too soon". If they could be reliably cleaned up when Mantid closes that would be great, but I'm not sure of the machinery for that.

adriazalvarez commented:
I see, thank you for the detailed explanation. 
Yes, maybe investigating how taxing is to create the cache files and how users are likely to implement scripts which would require reforming many cache files is outside the scope of this pr, but it could be interesting to investigate how to handle this differently. 
I have added a couple comments, then I think is good to go.

ajjackson commented on docs/source/algorithms/Abins-v1.rst:
```diff
@@ -61,7 +61,7 @@ Usage
 
     benzene_wrk = Abins(AbInitioProgram="CASTEP", VibrationalOrPhononFile="benzene.phonon",
-                        QuantumOrderEventsNumber="1")
+                        QuantumOrderEventsNumber="1", CacheDirectory="/tmp")
```
I'd much rather have nothing here at all, and use the default (User Directory). But for some reason it doesn't work; do you know if someone could advise on why this FileProperty (with a default!) shows up as a mandatory argument from the Python shell? Possibly I'm using it wrong, the docs were a bit C++-oriented

ajjackson replied:
Re: TemporaryDirectory, we _could_ create one and clean it up in this code with a context manager... but then it wouldn't be a very good doctest because it would look intimidating and not reflect real-world usage.

adriazalvarez replied:
I agree with being better not having any cache argument at all for the simple example. 
What do you with it not working? I'm running the doctests locally with the argument omitted and it works as long as I don't try to remove any nonexisting file in the test cleanup.

ajjackson replied:
Hmm, I'll take another look then!

ajjackson commented:
> Also as a general comment, maybe dragging the cache_directory parameter around few methods is a bit unsatisfactory. I've seen in other parts of mantid that temporary settings could be stored on mantid config files with a custom entry. We could write the temporary setting on the config file when calling the algorithm (or not write if it's already the same) and then retrieve locally in the methods that needs it. Would it be very inconvenient to do something like this ? Let me know what you think.

This is closer to how it worked before, using the Default Save Directory which is a global Mantid setting.

From a maintenance standpoint I have found this "magical" caching without explicit arguments to be quite confusing and awkward to debug. It isn't very obvious from the function names and signatures what will be cached, and unit tests could run into race conditions by working on the same cache file. This can also be an issue when running multiple parallel instances in an automated workflow such as https://github.com/ajjackson/abins-qpt-convergence/tree/main

adriazalvarez commented:
> I'm very keen to try the [joblib](https://joblib.readthedocs.io/en/stable/memory.html) approach: it is a bit like Python's [standard library memoization tool](https://docs.python.org/3/library/functools.html#functools.cache) but a) plays more nicely with complex objects especially numpy arrays; b) uses disk rather than memory. By memoizing private "pure" functions that take data as input and don't look at any quasi-global state, we should be able to drop a lot of the existing (and somewhat wasteful) logic that determines when the cache should be invalidated. Instead the cache is simply used whenever a function is called with exactly the same arguments (by hashing the data) as previously.

I'm not opposed to this, but you may have to talk to Sarah or Tom if you want to integrate this into the developer package of Mantid.

ajjackson commented:
It's already in the mantid dependencies, for quickBayes

ajjackson committed:
Doctests: default parameter for CacheDirectory seems to work!

It has been a bit of a headache getting these to work nicely, but it
may have been something to do with spacing between RST lines.

There seems to be a little API change at work here: the default cache
director is now always the default save directory. Previously, it
seems that command-line runs of Mantid would sometimes use the current
directory from which Mantid was launched.

That may be a headscratcher or even breaking for certain user scripts.
In such workflows it would be highly recommended to set an explicit
directory.

ajjackson committed:
Rework defaultsave / cleanup logic for Abins doctests

adriazalvarez commented on docs/source/algorithms/Abins-v1.rst:
```diff
@@ -58,6 +58,16 @@ Usage
 
 **Example - loading CASTEP phonon data:**
 
+.. testsetup:: AbinsCastepSimple
```
You may save some lines changing the setup to:
`.. testsetup:: *`
This applies the same setup to all groups .

ajjackson replied:
Very handy! Now that we are cleaning up a whole directory, the same cleanup command works for every test 😄

ajjackson committed:
Further Abins doctest cleanup

Use magic * to share test setup/cleanup logic across tests.
Move cleanup to top of file and use comments so it is easier to see
what is going on.

adriazalvarez reviewed (approved):
Thanks for the changes. 
Test are passing, and the handling of cache directories is more flexible and clear for users of Abins.
